### DISCUSSION TITLE
How to set timezone for default value when created column with default value now()?

### DISCUSSION THREAD

**Discussion by HwiLu** [ClickHouse/ClickHouse#60542] [General]
I mean there is a table like 
```sql
CREATE TABLE test_timezone
(
    `id` UInt32,
    `create_time` DateTime DEFAULT now(),
    `utc_time` DateTime DEFAULT now('UTC')
)
ENGINE = MergeTree
ORDER BY id
```
when I insert values, `now('UTC') ` seams does not work
```
┌─id─┬─────────create_time─┬────────────utc_time─┐
│  1 │ 2024-02-29 17:36:19 │ 2024-02-29 17:36:19 │
└────┴─────────────────────┴─────────────────────┘

```
my local timezome is `Asia/Shanghai`.

and I try to insert with `insert into test_timezone(id) SETTINGS  timezone='UTC'  values(1)`, but error happended  `Received exception from server (version 22.8.15):
Code: 115. DB::Exception: Received from localhost:9000. DB::Exception: Unknown setting timezone. (UNKNOWN_SETTING)`.

Is there any way to resolve it instead of modifying conf and restart ck server

**Comment by ArctypeZach** [comment]
Hi @HwiLu - looks like your clickhouse-server may be configured for UTC timezone.  You can adjust this in your config file: https://clickhouse.com/docs/en/operations/server-configuration-parameters/settings#timezone

Otherwise, maybe try:

```
CREATE TABLE test_timezone
(
    `id` UInt32,
    `create_time` DateTime DEFAULT now('Asia/Shanghai'),
    `utc_time` DateTime DEFAULT now('UTC')
)
ENGINE = MergeTree
ORDER BY id
```

**Reply by HwiLu** [reply]
Does not work, but still thank you. 😄

**Comment by den-crane** [comment]
>`create_time` DateTime DEFAULT now('Asia/Shanghai')

This has no sense. `now('Asia/Shanghai')` will be immediately converted to `DateTime` timezone. 
In other words `now()` returns 1709222884 for any timezone, then create_time = 1709222884 in the table.
Inside Clickhouse all DateTime values are stored as Unix timestamps in UTC timezone.
Timezone is a trait of DateTime type responsible for rendering and parsing DateTime text representation.

**Comment by den-crane** [comment]
>SETTINGS  timezone

setting is `session_timezone`

```
select now() settings session_timezone='Asia/Shanghai';
┌───────────────now()─┐
│ 2024-03-01 00:12:13 │
└─────────────────────┘

select now() settings session_timezone='UTC';
┌───────────────now()─┐
│ 2024-02-29 16:12:36 │
└─────────────────────┘
```

But `session_timezone` won't impact on numeric DateTimes, only on text.

```
create table X( D DateTime) Engine=Memory;
insert into X settings session_timezone='Asia/Shanghai' values('2000-01-01 00:00:00');
insert into X settings session_timezone='UTC' values('2000-01-01 00:00:00');

select * from X;
┌───────────────────D─┐
│ 2000-01-01 00:00:00 │
└─────────────────────┘
┌───────────────────D─┐
│ 2000-01-01 08:00:00 │
└─────────────────────┘
```
And in case `create table X(D DateTime('UTC') Engine=Memory;` UTC will supersede session_timezone.

**Reply by HwiLu** [reply]
All right, `session_timezone` supported since 23.6, my version is 22.8.15, I will try it later 😃
